### PR TITLE
Feature collecting citations to researcher's publications

### PR DESCRIPTION
**Pull request by ruybrito106**
Feature para requisitar ao Google Scholar o número de citações a uma dada publicação de um dado pesquisador. Nesse pull request estão implementados: cenários de controller, testes de controller, controllers, views e models.

---
This change is [Reviewable](https://reviewable.io/reviews/pauloborba/rpa/5)

### CODE REVIEW

File: test/cucumber/steps/CitationSteps.groovy [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Created by rbb3 on 01/11/16.
+ */
+
+
+
+this.metaClass.mixin(cucumber.api.groovy.Hooks)
+this.metaClass.mixin(cucumber.api.groovy.EN)
+
+Given(~/^the researcher "([^"]*)" with cpf "([^"]*)" is stored by the system$/) { String arg1, String arg2 ->
```
Comment (macm3) [inline test/cucumber/steps/CitationSteps.groovy]: arg1/arg2 não é intuitivo e não facilita a leitura

File: test/functional/pages/Citation.groovy [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import geb.Page
+
+/**
+ * Created by rbb3 on 03/11/16.
+ */
+class Citation extends Page {
+    static url = "/RPA/researcher/citations/"
+
+    static at = {
+        title ==~ /Ver Researcher/
```
Comment (macm3) [inline test/functional/pages/Citation.groovy]: teste não está internacionalizado

**Comment by pauloborba** [comment]
Reviewed 19 of 29 files at r1.
Review status: 19 of 29 files reviewed at latest revision, 12 unresolved discussions.

---

_[grails-app/controllers/rpa/ResearcherController.groovy, line 29 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsUUf55XcuG40MHLTN:-KWsUUf55XcuG40MHLTO:bdx6p5c) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/grails-app/controllers/rpa/ResearcherController.groovy#L29)):_

> ``` Groovy
>     def findCitations() {
>         Researcher res = Researcher.findByName(params.researcher)
>         Article art = Article.findByTittle(params.article)
> ```

devia pegar de todos os artigos do pesquisador, não?

---

_[grails-app/domain/rpa/Article.groovy, line 7 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsUhUPIrsPLo6DqKYD:-KWsUhUPIrsPLo6DqKYE:b1mou22) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/grails-app/domain/rpa/Article.groovy#L7)):_

> ``` Groovy
>     String journal
>     String issn
>     static hasMany = [authors:Researcher]
> ```

isso acima tem que ser discutido no canal do grupo, até porque está inconsistente com contribuições de outros membros. o mais adequado aqui seria Article estar associado a um conjunto de autores (String mesmo, já que eles podem ou não ser pesquisadores cadastrados no sistema). 

---

_[grails-app/domain/rpa/Article.groovy, line 8 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsV6e9oy487EK1tlNz:-KWsV6e9oy487EK1tlO-:b-lmozhw) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/grails-app/domain/rpa/Article.groovy#L8)):_

> ``` Groovy
>     String issn
>     static hasMany = [authors:Researcher]
>     static belongsTo = Researcher
> ```

isso é necessário mesmo? note também que, caso dois pesquisadores que tenham o mesmo artigo no lattes, podemos ter dois objetos artigos com a mesma informação, exceto os owners, certo? é isso mesmo que é esperado? 

---

_[test/cucumber/Citation.feature, line 11 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsWjT-y-CjKvGeN3bB:-KWsWjT2A7eVt5_8cAE6:b-3gxri9) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/cucumber/Citation.feature#L11)):_

> ``` Cucumber
> 
>     Given the researcher "Paulo Borba" with cpf "11111111111" is stored by the system
>     And the article with title "Implementing distribution and persistence aspects with AspectJ", issn "11111" from the journal "ACM" and with "376" citations at Google Scholar is stored by the system
> ```

o sistema não teria a quantidade de citações. isso teria que ser buscado no scholar. problema similar abaixo também

---

_[test/cucumber/steps/CitationSteps.groovy, line 36 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXCV1kHUS3d2PBS2u:-KWsXCV1kHUS3d2PBS2v:bk0vrb2) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/cucumber/steps/CitationSteps.groovy#L36)):_

> ``` Groovy
> }
> Then(~/^no citations from the article "([^"]*)" are stored by the system$/) { String arg1 ->
>     assert CitationTestSteps.informationStored(arg1, "0")
> ```

poucos cenários, nada de gui

---

_[test/functional/GebConfig.groovy, line 24 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXHxSferDZmyWrX5k:-KWsXHxTA08oVNjeehhA:b-835a6k) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/GebConfig.groovy#L24)):_

> ``` Groovy
> }
> 
> waiting {
> ```

por que precisou disso?

---

_[test/functional/pages/Citation.groovy, line 8 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXPfNlUr5oC9IMbUi:-KWsXPfNlUr5oC9IMbUj:b-plkmb7) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/pages/Citation.groovy#L8)):_

> ``` Groovy
>  * Created by rbb3 on 03/11/16.
>  */
> class Citation extends Page {
> ```

não entendi porque foi enviado se não ha nenhum cenário de gui

---

_[test/functional/steps/CitationTestSteps.groovy, line 13 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsYLNZ7qi0xw1SBaiT:-KWsYLNZ7qi0xw1SBaiU:buux35l) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/steps/CitationTestSteps.groovy#L13)):_

> ``` Groovy
> 
>     static public boolean compareCitationAmount(String article, String researcher) {
>         ArticleSteps articleSteps = new ArticleSteps()
> ```

não esta testando os servicos do sistema

---

_[test/functional/steps/CitationTestSteps.groovy, line 27 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXhm4j4udWPdgtNiS:-KWsXhm52oTJeowf4bBm:b2zhntb) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/steps/CitationTestSteps.groovy#L27)):_

> ``` Groovy
>     static public int findCitations(String article, String researcher) {
>         Article art = (new ArticleSteps()).findArticle(article)
>         GoogleScholarService gs = new GoogleScholarService()
> ```

tem que usar o controlador, o servico do sistema, sendo não estara testando 

---

_[test/functional/steps/CitationTestSteps.groovy, line 31 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXnlMqWZGBYR5-fVp:-KWsXnlMqWZGBYR5-fVq:b3kre) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/steps/CitationTestSteps.groovy#L31)):_

> ``` Groovy
>         list.add(art)
>         gs.findCitations(list)
>         System.out.println(list.get(0).citationAmount+"+")
> ```

?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)_

**Comment by ruybrito106** [comment]
*[grails-app/controllers/rpa/ResearcherController.groovy, line 29 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsUUf55XcuG40MHLTN:-KXzvcUL-BKib0RLUuvN:b-h5iml9) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/grails-app/controllers/rpa/ResearcherController.groovy#L29)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

devia pegar de todos os artigos do pesquisador, não?
</blockquote></details>

Eu ajeitei isso, professor. Adicionei a opção de buscar no google scholar pelo artigo ou pelo pesquisador. E os testes também validam essa nova opção.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[grails-app/domain/rpa/Article.groovy, line 7 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsUhUPIrsPLo6DqKYD:-KXzw70sDxomd2eOoYYM:b-f9sph6) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/grails-app/domain/rpa/Article.groovy#L7)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

isso acima tem que ser discutido no canal do grupo, até porque está inconsistente com contribuições de outros membros. o mais adequado aqui seria Article estar associado a um conjunto de autores (String mesmo, já que eles podem ou não ser pesquisadores cadastrados no sistema). 
</blockquote></details>

Já foi consertado. Conversei com outros membros e utilizei o mesmo template nas classes básicas.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[grails-app/domain/rpa/Article.groovy, line 8 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsV6e9oy487EK1tlNz:-KXzwFhgN8HrYSsfUQrj:b-f9sph6) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/grails-app/domain/rpa/Article.groovy#L8)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

isso é necessário mesmo? note também que, caso dois pesquisadores que tenham o mesmo artigo no lattes, podemos ter dois objetos artigos com a mesma informação, exceto os owners, certo? é isso mesmo que é esperado? 
</blockquote></details>

Já foi consertado. Conversei com outros membros e utilizei o mesmo template nas classes básicas.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/cucumber/Citation.feature, line 11 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsWjT-y-CjKvGeN3bB:-KXzwKSlHDV4BxWF6OHD:b-jtfq80) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/cucumber/Citation.feature#L11)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

o sistema não teria a quantidade de citações. isso teria que ser buscado no scholar. problema similar abaixo também
</blockquote></details>

Eu ajeitei isso também. O sistema só esta guardando o número de citações que é buscado no google shcolar. Os testes validam se o valor que é armazenado condiz com o valor que é buscado no scholar.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/cucumber/steps/CitationSteps.groovy, line 17 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KW0VfuAqPQ17nEDaAGd-r1-17:-KXzwiQ-3BBNWJ1nTKQ0:btot2yg) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/cucumber/steps/CitationSteps.groovy#L17)):*
<details><summary><i>Previously, macm3 (Marcela Azevedo) wrote…</i></summary><blockquote>

arg1/arg2 não é intuitivo e não facilita a leitura
</blockquote></details>

Já foi trocado.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/cucumber/steps/CitationSteps.groovy, line 36 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXCV1kHUS3d2PBS2u:-KXzwmNzj0iRfb6WRlzx:b-o48j54) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/cucumber/steps/CitationSteps.groovy#L36)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

poucos cenários, nada de gui
</blockquote></details>

Adicionei cenários de GUI.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/functional/GebConfig.groovy, line 24 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXHxSferDZmyWrX5k:-KXzwrtkLVjqcSsW7xuA:b2zahqw) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/GebConfig.groovy#L24)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

por que precisou disso?
</blockquote></details>

Por que os testes não estavam mockados e a resposta do servidor demorava um pouco. Por isso, para não quebrar os testes de controller eu usei o waiting.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/functional/steps/CitationTestSteps.groovy, line 13 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsYLNZ7qi0xw1SBaiT:-KXzxD82yBtPP5QuPNjZ:b-skjpx) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/steps/CitationTestSteps.groovy#L13)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

não esta testando os servicos do sistema
</blockquote></details>

Percebi isso. Os dados estavam estáticos no sistema e eu não estava de fato criando artigos ou pesquisadores. Mas isso foi consertado.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/functional/steps/CitationTestSteps.groovy, line 27 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXhm4j4udWPdgtNiS:-KXzxTVxR0SW-RsIj1-m:b-e9eibj) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/steps/CitationTestSteps.groovy#L27)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

tem que usar o controlador, o servico do sistema, sendo não estara testando 
</blockquote></details>

Percebi isso. Já foi consertado. Resolvei usando findBy(...) em atributos das classes básicas.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/functional/steps/CitationTestSteps.groovy, line 31 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXnlMqWZGBYR5-fVp:-KXzxguAPkib-pggHWWy:bem0aer) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/steps/CitationTestSteps.groovy#L31)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

?
</blockquote></details>

Foi sem querer. Esqueci de remover esse debug. Mas já removi.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/functional/pages/Citation.groovy, line 8 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KWsXPfNlUr5oC9IMbUi:-KXzxoENjdJNOG9mk6jn:b-auqof2) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/pages/Citation.groovy#L8)):*
<details><summary><i>Previously, pauloborba (Paulo Borba) wrote…</i></summary><blockquote>

não entendi porque foi enviado se não ha nenhum cenário de gui
</blockquote></details>

O senhor havia dito que era pra ser enviado o que tivesse sido feito. Desculpa, não entendi que não poderia ter enviado. Mas os cenários de gui já foram implementados pra segunda iteração.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

**Comment by ruybrito106** [comment]
*[test/functional/pages/Citation.groovy, line 12 at r1](https://reviewable.io:443/reviews/pauloborba/rpa/5#-KW0VfuAqPQ17nEDaAGe-r1-12:-KXzyEP8p5DJ578RWaTK:b-57rvxe) ([raw file](https://github.com/pauloborba/rpa/blob/b100b5e0d0a6343ecd4fc32a963b641de22c75fb/test/functional/pages/Citation.groovy#L12)):*
<details><summary><i>Previously, macm3 (Marcela Azevedo) wrote…</i></summary><blockquote>

teste não está internacionalizado 
</blockquote></details>

Isso foi feito pra segunda iteração.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/5)*

File: test/cucumber/steps/CitationSteps.groovy [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    }
+    def returnTitle = page.getTitleValue()
+    def value = page.getCitationValue()
+    assert returnTitle == title && value == citations
+}
+
+/*
+    Scenario: Irregular importing citations to article
+*/
+Then(~/^the message "([^"]*)" is displayed$/) { String message ->
```
Comment (LcsSantana) [inline test/cucumber/steps/CitationSteps.groovy:102]: Um step do cenário está com if e else. Nesse caso, provavelmente deveria haver dois cenários. E se tratando de um Then o comportamento deveria ser o mesmo para os mesmos Given e When.

File: grails-app/domain/rpa/Article.groovy [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    static constraints = {
+        title(nullable: false, blank: false)
+        journal(nullable: false, blank: false)
+        issn(nullable: false, blank: false)
+        owner(nullable: true)
+        citationAmount(blank: true)
+    }
+}
```
Comment (LcsSantana) [inline grails-app/domain/rpa/Article.groovy:23]: Não colocou a linha vazia no final.

File: test/cucumber/steps/CitationSteps.groovy [hunk 5]
```diff
@@ -0,0 +1,124 @@
+package steps
```
Comment (LcsSantana) [inline test/cucumber/steps/CitationSteps.groovy:1]: Está faltando o arquivo env.groovy para limpar os steps.

**Pull request closed without merging**